src/main/java/org/dependencytrack/persistence/ProjectQueryManager.java

```diff
@@ -864,6 +868,264 @@ public void recursivelyDelete(final Project project, final boolean commitIndex)
+                if (Config.getInstance().getProperty(Config.AlpineKey.DATABASE_DRIVER).equals("org.postgresql.Driver")) {
+                    sqlQuery = pm.newQuery(JDOQuery.SQL_QUERY_LANGUAGE, """
+                        WITH v_ids AS (
+                            SELECT "VIOLATIONANALYSIS"."ID"
+                            FROM "VIOLATIONANALYSIS"
+                            JOIN "PROJECT" ON "PROJECT_ID" = "PROJECT"."ID"
+                            WHERE "PROJECT"."UUID" = ANY(?)
+                        )
+                        DELETE FROM "VIOLATIONANALYSISCOMMENT"
+                        USING v_ids
+                        WHERE "VIOLATIONANALYSIS_ID" = v_ids."ID";
+                        """);
+                    executeAndCloseWithArray(sqlQuery, (Object) uuidsArray);
+                }
+
+                sqlQuery = pm.newQuery(JDOQuery.SQL_QUERY_LANGUAGE, """
+                    DELETE FROM "VIOLATIONANALYSIS" WHERE "PROJECT_ID" IN (
+                        SELECT "ID" FROM "PROJECT" WHERE "UUID" = ANY(?)
+                    );
+                    """);
+                executeAndCloseWithArray(sqlQuery, (Object) uuidsArray);
+
+                sqlQuery = pm.newQuery(JDOQuery.SQL_QUERY_LANGUAGE, """
+                    DELETE FROM "POLICYVIOLATION" WHERE "PROJECT_ID" IN (
+                        SELECT "ID" FROM "PROJECT" WHERE "UUID" = ANY(?)
+                    );
+                    """);
+                executeAndCloseWithArray(sqlQuery, (Object) uuidsArray);
+
+                sqlQuery = pm.newQuery(JDOQuery.SQL_QUERY_LANGUAGE, """
+                    DELETE FROM "POLICY_PROJECTS" WHERE "PROJECT_ID" IN (
+                        SELECT "ID" FROM "PROJECT" WHERE "UUID" = ANY(?)
+                    );
+                    """);
+                executeAndCloseWithArray(sqlQuery, (Object) uuidsArray);
+
+                sqlQuery = pm.newQuery(JDOQuery.SQL_QUERY_LANGUAGE, """
+                    DELETE FROM "POLICY"
+                    WHERE "ID" NOT IN (SELECT "POLICY_ID" FROM "POLICY_PROJECTS");
+                    """);
+                executeAndCloseWithArray(sqlQuery, (Object) uuidsArray);
+
+                sqlQuery = pm.newQuery(JDOQuery.SQL_QUERY_LANGUAGE, """
+                    DELETE FROM "PROJECT_ACCESS_TEAMS" WHERE "PROJECT_ID" IN (
+                        SELECT "ID" FROM "PROJECT" WHERE "UUID" = ANY(?)
+                    );
+                    """);
+                executeAndCloseWithArray(sqlQuery, (Object) uuidsArray);
+
+                sqlQuery = pm.newQuery(JDOQuery.SQL_QUERY_LANGUAGE, """
+                    DELETE FROM "VEX" WHERE "PROJECT_ID" IN (
+                        SELECT "ID" FROM "PROJECT" WHERE "UUID" = ANY(?)
+                    );
+                    """);
+                executeAndCloseWithArray(sqlQuery, (Object) uuidsArray);
+
+                // The below has only been tested with Postgres, but should work on any RDBMS supporting SQL:1999
```

Comment:
For completeness, how many RDBMSes have you tested with this? Having tested at least H2, MSSQL, and PostgreSQL would be good. We have a Docker Compose file for MSSQL here: https://github.com/DependencyTrack/dependency-track/blob/master/dev/docker-compose.mssql.yml

Reply:
I've only tested it with H2 and PostgreSQL so far (H2 in unit tests, PostgreSQL in integration tests) but I can try migrating my test data to MSSQL and test that as well. I'll get back to you with the results when done!

Reply:
No need to do a full load test or anything. We just need assurance that the queries work as expected. Sadly the small differences in SQL syntax between the different RDBMSes turn out to be tripwires more often than not. :D

Reply:
True, and true :) But I'll need some data in the schema to pass the condition that the UUID list is not empty, so I'll try to migrate the relatively small exported sample that I created with [pg_sample](https://github.com/mla/pg_sample).